### PR TITLE
Do not qualify instance member access within a constructor initializer

### LINKED ISSUE #32093: Style preferences to include `this.` don't work in ctor
A constructor calling its base class' constructor can't use `this.`, but auto-formatting applies the chage.

```csharp
public class Base{
  public string Name{get;}
  public Base(string name){...}
}

public class Child: Base{  
  public Child():
     base(AppContext.GetValueFor(nameof(this.Name)){
}
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/415902/style-preferences-to-include-this-dont-work-in-cto.html
VSTS ticketId: 757067_
_These are the original issue comments:_

Jinu Joseph [MSFT] on 1/1/2019, 05:31 PM (17 hours ago): <p>We appreciate you taking the time to report this problem. We are currently prioritizing problems that are impacting a broad set of our customers, so we may not be able to investigate this one immediately. We know this problem is important to you, so we will continue to monitor it.</p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by KGRWhite**
Fixes #32093

### CODE REVIEW

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -21,14 +21,20 @@ protected override bool IsAlreadyQualifiedMemberAccess(ExpressionSyntax node)
...
+        // or member in property or field initialization,
+        // or member in base constructor,  it cannot be qualified.
         protected override bool CanMemberAccessBeQualified(ISymbol containingSymbol, SyntaxNode node)
         {
             if (node.GetAncestorOrThis<AttributeSyntax>() != null)
             {
                 return false;
             }
 
+            if (node.GetAncestorsOrThis(n => n.IsKind(SyntaxKind.BaseConstructorInitializer) && n.Contains(node)).Any())
```
Comment (sharwell) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: 💭 Is there any reason why we can't simplify this to the following?

```csharp
if (node.GetAncestorOrThis<ConstructorInitializerSyntax>() != null)
```

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -21,14 +21,20 @@ protected override bool IsAlreadyQualifiedMemberAccess(ExpressionSyntax node)
 
         // If the member is already qualified with `base.`,
         // or member is in object initialization context,
-        // or member in property or field initialization, it cannot be qualified.
+        // or member in property or field initialization,
+        // or member in base constructor,  it cannot be qualified.
```
Comment (sharwell) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]:
```suggestion
        // or member in constructor initializer, it cannot be qualified.
```

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs (hunk 1 above)
Comment (KGRWhite) [reply src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: No reason other then I hadn't realized I could simplify it at the time :-)
Please see new commit which addresses this.

File: src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs [hunk 3]
```diff
@@ -1373,7 +1373,25 @@ public async Task DoNotReportToQualify_IfInStaticContext6()
 }",
 CodeStyleOptions.QualifyEventAccess);
         }
-
+		[WorkItem(32093, "https://github.com/dotnet/roslyn/issues/32093")]
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs]:
```suggestion

        [WorkItem(32093, "https://github.com/dotnet/roslyn/issues/32093")]
```

File: src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs [hunk 4]
```diff
@@ -1373,7 +1373,25 @@ public async Task DoNotReportToQualify_IfInStaticContext6()
...
+    public Base(string foo){}
+}
+public class Derived : Base
+{
+    public Derived()
+        : base(nameof([|Foo|]))
+    {}
+}
+",
+  CodeStyleOptions.QualifyFieldAccess);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs]:
```suggestion
                CodeStyleOptions.QualifyFieldAccess);
```

File: src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs [hunk 5]
```diff
@@ -1373,7 +1373,25 @@ public async Task DoNotReportToQualify_IfInStaticContext6()
...
+        }
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/QualifyMemberAccess/QualifyMemberAccessTests.cs]:
```suggestion
        }

```

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 6]
```diff
@@ -21,14 +21,20 @@ protected override bool IsAlreadyQualifiedMemberAccess(ExpressionSyntax node)
...
         {
             if (node.GetAncestorOrThis<AttributeSyntax>() != null)
             {
                 return false;
             }
 
+            if (node.GetAncestorsOrThis<ConstructorInitializerSyntax>() != null)
+            {
+                return false;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: Do we need to make (or at least test for) a similar issue on the VB side?

Comment (KGRWhite) [reply src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: I'll investigate and rectify if required. My VB is a little bit rusty though :-)

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: no worries.  in VB you'd do: 

```vb
sub new(...)
   me.new(...)  ' or mybase.new(...)
end sub
```

Comment (KGRWhite) [reply src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: I cant seem to replicate the issue in VB and the new test seems to pass (latest commit)

**Comment by sharwell** [comment]
@KGRWhite Thanks for updating this! I'm not sure why the build is failing so I will investigate and see about getting this merged 👍

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -21,14 +21,20 @@ protected override bool IsAlreadyQualifiedMemberAccess(ExpressionSyntax node)
...
+        // or member in property or field initialization,
+        // or member in constructor initializer, it cannot be qualified.
         protected override bool CanMemberAccessBeQualified(ISymbol containingSymbol, SyntaxNode node)
         {
             if (node.GetAncestorOrThis<AttributeSyntax>() != null)
             {
                 return false;
             }
 
+            if (node.GetAncestorsOrThis<ConstructorInitializerSyntax>() != null)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: :exclamation: So the issue here is super subtle, but the fix is easy. The method used here, `GetAncestorsOrThis` returns an `IEnumerable<T>` which can be empty but is never null. What we actually wanted is the equivalent of `GetAncestorsOrThis<T>().FirstOrDefault()`, which it turns out is available under another name which differs from the current code by only one letter:

```suggestion
            if (node.GetAncestorOrThis<ConstructorInitializerSyntax>() != null)
```

**Comment by sharwell** [comment]
@KGRWhite thank you again 👍

**Pull request merged**